feat(rate-limiting-advanced): Throttle requests (#2697)
Kong/developer.konghq.com: app/_kong_plugins/rate-limiting-advanced/index.md

  - q: What happens to queued requests if a client drops its connection with {{site.base_gateway}} during the Rate Limiting Advanced throttling period?
    a: |
      If a client drops its connection with Kong while a request is being throttled ({% new_in 3.12 %}), {{site.base_gateway}} automatically releases all associated resources for that specific request. This means the individual request will no longer be processed or retried. However, the counter that accounted for this request's slot in the "waiting room" is automatically managed by the underlying counter mechanism (shared dictionary or Redis). These counters are typically recorded within specific time windows and are automatically evicted when their window expires, ensuring resource cleanup without manual intervention for each dropped connection.

> ---

Reviewer: you could use the following docker image, which i got it from latest commits in github.com/kong-ee/kong-ee yesterday
```
kong/kong-gateway-dev:548625e3ebfb9f8ee615e30d57a26c16b855dd14-ubuntu
```

> ---

Reviewer: I setup a konnect CP env in konnect dev: https://cloud.konghq.tech/eu/gateway-manager/c3f9e524-5150-4821-bd4c-7bbf66067d2a/overview. you could use this. Or you could setup a throttling config by yourself

<img width="1544" height="1052" alt="image" src="https://github.com/user-attachments/assets/6341a3bb-3a31-43fa-aeb0-9225f520c391" />

min_version:
  gateway: '1.0'
faqs:
  - q: What are the potential impacts and risks associated with enabling request throttling in Rate Limiting Advanced?
    a: |

>       Enabling request throttling ({% new_in 3.12 %}) can lead to a degradation in the capacity of {{site.base_gateway}} data plane nodes. This is because client requests are held open for a longer duration during the throttling period compared to normal rejections. This extended occupation of resources (like memory and file descriptors) can reduce the data plane's ability to handle other new requests, potentially leading to scale or stress issues during high traffic spikes. Configuring a large `config.throttle_queue_limit` can also consume significant memory on data plane nodes.

Reviewer 2: ```suggestion
      Enabling request throttling can lead to a degradation in the capacity of {{site.base_gateway}} data plane nodes. This is because client requests are held open for a longer duration during the throttling period compared to normal rejections. This extended occupation of resources (like memory and file descriptors) can reduce the data plane's ability to handle other new requests, potentially leading to scale or stress issues during high traffic spikes. Configuring a large [`config.throttle_queue_limit`](/plugins/rate-limiting-advanced/reference/#schema--config-throttle-queue-limit) can also consume significant memory on data plane nodes.
```
What do you think about not including the "new_in" in these FAQs? I'm wondering if there's a limit for this, and if we're overloading the docs by having to new_in a feature in every single mention. It makes sense for the example and for the section on the overview, but including it in FAQs feels like overkill.

  - q: What happens to queued requests if a client drops its connection with {{site.base_gateway}} during the Rate Limiting Advanced throttling period?
    a: |
      If a client drops its connection with Kong while a request is being throttled ({% new_in 3.12 %}), {{site.base_gateway}} automatically releases all associated resources for that specific request. This means the individual request will no longer be processed or retried. However, the counter that accounted for this request's slot in the "waiting room" is automatically managed by the underlying counter mechanism (shared dictionary or Redis). These counters are typically recorded within specific time windows and are automatically evicted when their window expires, ensuring resource cleanup without manual intervention for each dropped connection.
  - q: How is memory usage impacted when I enable throttling with the Rate Limiting Advanced plugin?
    a: |

>       ({% new_in 3.12 %}) In extreme conditions where both {{site.base_gateways}}’s header buffer and the kernel’s TCP buffer are fully used and you're using the default configuration ({{site.base_gateways}} accepts a maximum request header size of 32 KB, and the Linux kernel TCP buffer is approximately 200 KB), the average memory consumption of each open connection is around 220 KB for one Route with one Rate Limiting Advanced plugin configured with the following:

Reviewer 2: ```suggestion
      In regular conditions, memory usage is minimally impacted. In extreme conditions where both {{site.base_gateways}}’s header buffer and the kernel’s TCP buffer are fully used and you're using the default configuration ({{site.base_gateways}} accepts a maximum request header size of 32 KB, and the Linux kernel TCP buffer is approximately 200 KB), the average memory consumption of each open connection is around 220 KB for one Route with one Rate Limiting Advanced plugin configured with the following:
```
Adding a "in regular conditions" sentence to make it less alarming.

subsets of Consumers.

You can see an example of this in the guide on [enforcing rate limiting tiers with the Rate Limiting Advanced plugin](/how-to/add-rate-limiting-tiers-with-kong-gateway/).

## Throttle rate limits {% new_in 3.12 %}

> In {{site.base_gateway}} 3.12, you can enable request throttling via the Rate Limiting Advanced plugin to improve clients' experience and protect upstream origin servers from being overwhelmed by traffic spikes. By throttling requests, this allows requests that exceed the rate limit threshold to be delayed and retried, rather than immediately rejected with a `429` status code. 

Reviewer 2: ```suggestion
In {{site.base_gateway}} 3.12 or later, you can enable request throttling using the Rate Limiting Advanced plugin to improve clients' experience and protect upstream origin servers from being overwhelmed by traffic spikes. With throttling, requests that exceed the rate limit threshold can be delayed and retried, rather than immediately rejected with a `429` status code. 
```

Throttled rate limits work like the following:

> 1. When a request hits the rate limit, it is placed into a "waiting room" or queue. The client's connection is held during this delay.

Reviewer 2: ```suggestion
1. When a request hits the rate limit, it's placed into a "waiting room" or queue. The client's connection is held during this delay.
```

   * This queue uses local, Redis, or cluster strategies to manage the queue of throttled requests using a counter-based approach.

> 1. Requests in the queue are automatically retried after a configurable interval (`config.interval`). 
>    * There's a limit to retries for individual requests (`config.retry_times`), and a cap to the total number of requests waiting (`config.queue_limit`).
>    * All concurrent requests will retry at approximately the same time once the specified interval has elapsed.

Reviewer 2: ```suggestion
1. Requests in the queue are automatically retried after a configurable interval ([`config.interval`](/plugins/rate-limiting-advanced/reference/#schema--config-interval)). 
   * There's a limit to retries for individual requests ([`config.retry_times`](/plugins/rate-limiting-advanced/reference/#schema--config-retry-times)), and a cap to the total number of requests waiting ([`config.queue_limit`](/plugins/rate-limiting-advanced/reference/#schema--config-queue-limit)).
   * All concurrent requests will retry at approximately the same time once the specified interval has elapsed.
```
